### PR TITLE
FINERACT-2090: Restructure Withdraw Loan Application Logic

### PR DESCRIPTION
**Pull request by ruchiD**
## Description
Restructure Withdraw Loan Application Logic
Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java [hunk 1]
```diff
@@ -853,34 +854,21 @@ private Map<String, Object> loanApplicationRejection(Loan loan, final JsonComman
...
-        final AppUser currentUser = getAppUserIfPresent();
-
-        this.loanApplicationTransitionValidator.validateApplicantWithdrawal(command.json());
-
+        // retrieve loan
         final Loan loan = retrieveLoanBy(loanId);
-        loanApplicationTransitionValidator.checkClientOrGroupActive(loan);
 
+        // validate withdrawal
+        loanApplicationTransitionValidator.validateApplicantWithdrawal(command, loan, defaultLoanLifecycleStateMachine);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanApplicationWritePlatformServiceJpaRepositoryImpl.java]: You dont need to send the defaultLoanLifecycleStateMachine as a parameter. You can inject in the loanApplicationTransitionValidator directly

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationTransitionValidator.java [hunk 2]
```diff
@@ -237,6 +283,14 @@ public void validateAccountStatus(final Loan loan, final LoanEvent event) {
                     dataValidationErrors.add(error);
                 }
             }
+            case LOAN_WITHDRAWN -> {
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationTransitionValidator.java]: Is this used anywhere?

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationTransitionValidator.java [hunk 3]
```diff
@@ -204,6 +213,38 @@ public void validateApplicantWithdrawal(final String json) {
...
+            }
+
+            validateActivityNotBeforeClientOrGroupTransferDate(loan, LoanEvent.LOAN_WITHDRAWN, withdrawnOn);
+
+            if (DateUtils.isDateInTheFuture(withdrawnOn)) {
+                final String errorMessage = "The date on which a loan is withdrawn cannot be in the future.";
+                throw new InvalidLoanStateTransitionException("withdraw", "cannot.be.a.future.date", errorMessage, command);
+            }
+        } else {
+            final String errorMessage = "Only the loan applications with status 'Submitted and pending approval' are allowed to be withdrawn by applicant.";
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationTransitionValidator.java:236]: This validation and the one on line 286 seems duplicate. Do we need both?

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationTransitionValidator.java (hunk 2 above)
Comment (ruchiD) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanApplicationTransitionValidator.java]: Will remove this from method. It is not used anywhere. This is taken care with status transition check (line 242).

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**